Revise presentation of feature-state shortcode

lmktfy opened the pull request:
Handle removed features differently. When a feature has graduated to stable and has since been removed, don't show the info box that we use for feature gates that Kubernetes still recognizes. Instead, show a simple paragraph explaining the situation.

/area web-development

The idea I have is to use the metadata we already have (`removed`) to decide what text to show to readers.

For previews, see https://github.com/kubernetes/website/pull/52632#issuecomment-3369232954

lmktfy committed:
Revise shortcode for feature gates

Handle removed features differently. When a feature has graduated to
stable and has since been removed, don't show the info box that we
use for feature gates that Kubernetes still recognizes. Instead, show
a simple paragraph explaining the situation.

lmktfy commented:
Examples of feature gates that are not yet stable:

https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/compositepodgroup-api/

https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/concepts/resource-management/dynamic-resource-allocation/dra-features/#partitionable-devices

---

Example of a feature gate _not_ marked as removed, legacy means to specify it:
https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/reference/access-authn-authz/authentication/#bootstrap-tokens

_No explanatory text_ but still works.

---

Example of a feature gate _not_ marked as removed, recommended means to specify it, recent graduation:
https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/reference/access-authn-authz/authentication/#using-authentication-configuration

> This is a stable feature in Kubernetes, and has been since version 1.34. It was first available in the v1.29 release.

---

Example of a feature gate **marked as removed**, recommended means to specify it:
https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/concepts/architecture/nodes/#node-topology

> This is a stable feature in Kubernetes, and has been since version 1.27. It was first available in the v1.16 release. You can no longer disable or opt of of this behavior.

I have deliberately picked an example that is a little stale.

Even if we make this this change, we should revise https://kubernetes.io/docs/concepts/architecture/nodes/#node-topology to remove ”_If you have enabled the TopologyManager [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/), then&hellip;_”. I would also remove the feature gate shortcode from that section at the same time, as it's been stable for a while. However, this PR is not the place to fix that. We could use https://github.com/kubernetes/website/pull/52626 for that work, or tweak https://github.com/kubernetes/website/pull/52626 to leave in some feature gates where the removal is relatively recent.

Bear in mind that if we add metadata to record whether feature gates are locked, we can tweak the logic without needing to change the message. The statements in the docs would remain true.

lmktfy commented:
Want different text? Feel free to suggest changes to it.

tengqm commented:
There are some slight differences between the following states of a feature gate:

- GA'ed but not "locked": you can specify the gate, and you can opt out;
- GA'ed and "locked": you can specify the gate, but you cannot opt out;
- Removed: you can NOT specify the gate, it is completely gone, no longer recognized as a valid gate.

lmktfy commented:
> There are some slight differences between the following states of a feature gate:
> 
> * GA'ed but not "locked": you can specify the gate, and you can opt out;
> * GA'ed and "locked": you can specify the gate, but you cannot opt out;
> * Removed: you can NOT specify the gate, it is completely gone, no longer recognized as a valid gate.

I agree. And I hope that a future PR continues our improvement journey.
 
> For gates that have been **removed**, we may want to drop the anchors instead of carry them on forever.

Anchors?

If you mean shortcodes, then removing the shortcodes can still happen. I propose retaining the shortcode for a number of releases after it has been removed, or since it became stable.

lmktfy committed:
Expand feature gate shortcode

* When not enabled by default, append an expandable <details> that allows
  revealing extra docs including a link to the relevant task page
* When graduated and stable, say so.
* When graduated and stable and removed, simplify the message.
* When graduated and stable and removed but OFF by default, customize
  even further for this specific case.

lmktfy commented:
@SayakMukhopadhyay or @reylejano would you take a look and let me know your feedback?

Caesarsage commented on layouts/shortcodes/feature-state.html:
```diff
@@ -1,52 +1,169 @@
+            {{- end -}}
+
+            {{/* Handle removed feature gates */}}
+            {{- if and (eq .stage "stable") ($featureGateWasRemoved) .defaultValue -}}
```
Can we handle `removed` before checking for `stable`? Some removed gates never reached stable. For example, `DRAControlPlaneController` is `removed: true` with final stage `alpha`, so this condition lets it fall through to the “enable this feature gate” guidance.

lmktfy replied:
I revised the logic here.

Caesarsage commented on layouts/shortcodes/feature-state.html:
```diff
@@ -1,52 +1,169 @@
+              {{/* Special message for feature that are both stable ann disabled by default */}}
+              {{ if not .defaultValue -}}
+                <em>{{ T "feature_gate_stable_graduated_disabled" }}</em></p>
+              {{ end -}}
```
This `<p>` only closes when `.defaultValue` is false. you could move the closing `</p>` outside the conditional so stable/default-enabled gates render valid HTML too?

```suggestion
              <p><em>{{ T "feature_gate_stable_graduated" (dict "firstReleaseWithThisFeature" $firstReleaseWithThisFeature "stableVersion" .fromVersion ) | safeHTML }}</em>
              {{/* Special message for feature that are both stable ann disabled by default */}}
              {{ if not .defaultValue -}}
                <em>{{ T "feature_gate_stable_graduated_disabled" }}</em>
              {{ end -}}</p>
```

Caesarsage commented on layouts/shortcodes/feature-state.html:
```diff
@@ -1,52 +1,169 @@
+                <details>
+                <summary>{{ T "feature_gate_more_details" }}</summary>
+                {{ T "feature_gate_more_information_how_to_enable" (dict "feature_gate_name" $feature_gate_name ) | safeHTML }}
+                </p>
```
This looks like an extra closing `</p>` because `feature_gate_more_information_how_to_enable` already renders complete paragraph tags on the localized text here 

https://github.com/kubernetes/website/pull/52632/changes#diff-5ff7a9f4c0b97bf365afd8226631fe7bb7d70fcb6ed1b52f72dc98b15228ac98R212

```suggestion
         
```

Caesarsage reviewed (commented):
I found a few issues worth fixing

lmktfy commented:
I updated the PR. How's this looking now?

lmktfy commented:
I think this might close https://github.com/kubernetes/website/issues/49388

Caesarsage commented on layouts/shortcodes/feature-state.html:
```diff
@@ -1,52 +1,180 @@
+                 {{ T "feature_gate_details_partial" (dict "enabled_or_disabled" $featureGateEnabledOrDisabled "stage" (T (printf "feature_gate_stage_%s" .stage)) "kubernetes" (T "feature_state_kubernetes_label") "version" $for_k8s_version) | safeHTML }}
+               {{ else -}}
+                 {{/* Otherwise, show stage and enabled/disabled status */}}
+                 {{ T "feature_gate_details" (dict "enabled_or_disabled" $featureGateEnabledOrDisabled "stage" (T (printf "feature_gate_stage_%s" .stage)) "kubernetes" (T "feature_state_kubernetes_label") "version" $for_k8s_version) | safeHTML }}
```
Automatic mode passes `$for_k8s_version` here. Nothing supplies `for_k8s_version` in automatic mode, so line 23 falls back to `.Page.Param "version"`, the site's current release. The old code used `v{{ .fromVersion }}`, so every auto-mode box now states v1.36 regardless of when the stage actually began.

On the preview: 

1. ClusterTrustBundle (beta since 1.33) renders "Beta since Kubernetes v1.36"; 

https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/reference/access-authn-authz/certificate-signing-requests/#cluster-trust-bundles

2. ContainerStopSignals (alpha since 1.33) renders "Alpha since Kubernetes v1.36"; 

https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/pods/pod-lifecycle/#defining-custom-stop-signals

and

3. [DynamicResourceAllocation](https://deploy-preview-52632--kubernetes-io-main-staging.netlify.app/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) contradicts itself — the box says "Stable since Kubernetes v1.36" while its own details say "has been since version 1.35".

<img width="932" height="177" alt="Image" src="https://github.com/user-attachments/assets/9f2a55a5-ea9d-4db9-b6ce-496e4868ceec" />

lmktfy replied:
Eesh, sorry this was so wrong.

Caesarsage commented on i18n/en/en.toml:
```diff
@@ -192,11 +192,47 @@ other = "Error"
+other = """This is a stable feature in Kubernetes, and has been since the {{ .stableVersion }} release. You can no longer toggle this feature (the associated <a href="/docs/reference/command-line-tools-reference/feature-gates-removed/">feature gate</a> has been removed)."""
+
+[feature_gate_removed_before_ga_final_version_clear]
+other = """Kubernetes no longer includes this feature / behavior. The last release of Kubernetes that recognized the <tt>{{ .feature_gate_name }}</tt> feature gate was {{ .toVersion }}. You can no longer toggle that setting."""
```
This reads 'feature_gate_name' but the template passes featureGateName. 

```suggestion
other = """Kubernetes no longer includes this feature / behavior. The last release of Kubernetes that recognized the <tt>{{ .featureGateName }}</tt> feature gate was {{ .toVersion }}. You can no longer toggle that setting."""
```

lmktfy replied:
Fixed, I hope. Sorry about the bugs.

Caesarsage commented on layouts/shortcodes/feature-state.html:
```diff
@@ -1,52 +1,180 @@
               {{- end -}}
+              
+              {{/* Feature gate info box (automatic mode) */}}
+            <div class="feature-state-notice feature-{{ .stage }}{{ if and (eq .stage "stable") (not .defaultValue) }}feature-stable-but-disabled{{ end }}" title="{{ printf "%s %s" (T "feature_state_feature_gate_tooltip") $feature_gate_name }}">
```
Missing space, so the classes and SCSS rules will apply

```suggestion
            <div class="feature-state-notice feature-{{ .stage }}{{ if and (eq .stage "stable") (not .defaultValue) }} feature-stable-but-disabled{{ end }}" title="{{ printf "%s %s" (T "feature_state_feature_gate_tooltip") $feature_gate_name }}">
```

lmktfy replied:
Fixed, I hope. Sorry about the bugs.

Caesarsage commented:
> I think this might close #49388

Yes, it is close, but I don't think it fully closes #49388 yet. the `feature_gate_stable_locked` (https://github.com/kubernetes/website/pull/52632/changes#diff-5ff7a9f4c0b97bf365afd8226631fe7bb7d70fcb6ed1b52f72dc98b15228ac98R224) here  is defined but not referenced, so the `locked` metadata is read and never shown to readers.

Wiring this string into the stable branch would do it.

lmktfy commented:
How's this looking?

lmktfy commented:
@SayakMukhopadhyay would you, or anyone else from @kubernetes/website-maintainers, be willing to review this?

lmktfy commented:
D'oh, I found a snag. Will revise.

lmktfy commented:
Updated, and ready to review. Sorry about the noise.

Caesarsage commented:
Thank you, 
This definitely improves the feature-state shortcode

/lgtm
